mesonbuild/environment.py

```diff
@@ -1069,10 +1082,69 @@ def need_exe_wrapper(self):
             return False
         return True
 
-
 class MachineInfo:
     def __init__(self, system, cpu_family, cpu, endian):
         self.system = system
         self.cpu_family = cpu_family
         self.cpu = cpu
         self.endian = endian
+
+    @staticmethod
+    def detect(compilers = None):
+        """Detect the machine we're running on
+
+        If compilers are not provided, we cannot know as much. None out those
+        fields to avoid accidentally depending on partial knowledge. The
+        underlying ''detect_*'' method can be called to explicitly use the
+        partial information.
+        """
+        return MachineInfo(
+            detect_system(),
+            detect_cpu_family(compilers) if compilers is not None else None,
+            detect_cpu(compilers) if compilers is not None else None,
+            sys.byteorder)
+
+    @staticmethod
+    def from_literal(literal):
+        minimum_literal = {'cpu', 'cpu_family', 'endian', 'system'}
+        if set(literal) < minimum_literal:
+            raise EnvironmentException(
+                'Machine info is currently {}\n'.format(literal) +
+                'but is missing {}.'.format(minimum_literal - set(literal)))
+        return MachineInfo(
+            literal['system'],
+            literal['cpu_family'],
+            literal['cpu'],
+            literal['endian'])
+
+class MachineInfos:
+    def __init__(self):
+        self.build = None
+        self.host = None
+        self.target = None
+
+    def slide_missing(self):
```

Comment:
Why is this called `slide_missing`?

Reply:
Happy to change it. `shift_nonnull`? `default_to_previous`?

`slide_missing` is indeed a bad name as though I do think of "sliding" the definitions over, it is the non-missing ones, rather than missing ones, that "move".